rfcs/0004-persistent-ledger-builder-controller.md

```diff
@@ -0,0 +1,67 @@
+# Summary
+[summary]: #summary
+
+This adds a persistent backend for the ledger builder controller.
+
+# Motivation
+[motivation]: #motivation
+
+Having a persistent backend will allow nodes that go offline, whether for a short period or an extended period of time, to catch back up to the network with as little work as reasonably possible.
+
+# Detailed design
+[detailed-design]: #detailed-design
+
+In order to achieve this, we will persist the locked tip using the existing Merkle database. The backend Merkle database is designed in a way such that queries for the most commonly performed operations can be batched together as one request to the database.
+
+The best tip, materialized by the ledger builder controller, will be represented as a sparse Merkle ledger mask on top of the persistent Merkle ledger database. This data structure will only store the modified nodes of the Merkle ledger in memory, and defer to the contents of the persistent Merkle ledger database for any information outside of that. The persistent Merkle ledger database will maintain a reference to all of these data structures derived for it. These references will be used to garbage collect outdated information in the sparse diff representations whenever changes are written to the persistent copy. In other words, information will be removed from the sparse diff whenever it is no longer different from the information in the underlying database, ensuring that the data structure will not leak memory. If a sparse diff is empty, it may need to be retained, in case there are existing references to it.
```

Comment:
Suggested change: "whenever it is no longer different from" to "when it duplicates"